core/domain/prod_validation_jobs_one_off.py

```diff
@@ -15,29 +15,168 @@
 # limitations under the License.
 
 """One-off jobs for validating prod models."""
+
+import collections
+import copy
+import csv
+import datetime
+import re
+
+from constants import constants
 from core import jobs
+from core.domain import activity_domain
+from core.domain import collection_services
+from core.domain import commit_commands_domain
+from core.domain import exp_services
+from core.domain import recommendations_services
+from core.domain import story_services
 from core.platform import models
+import feconf
+import schema_utils
+import utils
 
-(user_models, exp_models, collection_models, feedback_models) = (
-    models.Registry.import_models([
-        models.NAMES.user, models.NAMES.exploration,
-        models.NAMES.collection, models.NAMES.feedback]))
+(
+    activity_models, audit_models, base_models,
+    collection_models, config_models, email_models,
+    exp_models, feedback_models, file_models,
+    recommendations_models, story_models,
+    user_models,) = (
+        models.Registry.import_models([
+            models.NAMES.activity, models.NAMES.audit, models.NAMES.base_model,
+            models.NAMES.collection, models.NAMES.config, models.NAMES.email,
+            models.NAMES.exploration, models.NAMES.feedback, models.NAMES.file,
+            models.NAMES.recommendations, models.NAMES.story,
+            models.NAMES.user]))
 datastore_services = models.Registry.import_datastore_services()
 
 
 class BaseModelValidator(object):
     """Base class for validating models."""
 
-    errors = {}
+    errors = collections.defaultdict(list)
     # external_models is keyed by field name. Each value consists
     # of the model class and a list of (external_key, external_model_instance)
     # tuples.
     external_models = {}
+    is_commit_log_entry_model = False
+    is_snapshot_metadata_model = False
+
+    @classmethod
+    def _get_model_id_regex(cls, item):
+        """Defines a regex for model id.
+
+        This should be implemented by subclasses.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+
+        Returns:
+            str. A regex pattern to be followed by the model id.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def _validate_model_id(cls, item):
+        """Checks whether the id of model matches the regex specified for
+        the model.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+        """
+        regex_string = cls._get_model_id_regex(item)
+        if not re.compile(regex_string).match(str(item.id)):
+            cls.errors['model id check'].append((
+                'Entity id %s: Entity id does not match regex pattern') % (
+                    item.id))
+
+    @classmethod
+    def _get_json_properties_schema(cls, item):
+        """Defines a schema for model properties.
```

Comment:
This is not mandatory for all properties due to two reasons:

- Validation for some of the json properties is already present in the validation of model domain object class.
- Some properties like contributor_summaries in StorySummaryModel do not have a fixed schema since the keys are user_ids of contributors, so we cannot have a particular name defined for each key

I now think that a clean way to do this check would be to remove this function and add validation for missing json properties in model domain object class validation function. What do you think?

Reply:
SGTM in general. Though, just to clarify, what do you mean by "missing json properties"?

Reply:
I meant the json properties which are not validated in the validate function of the domain object class.

Reply:
Done.